forge-ai/src/main/java/forge/ai/simulation/SpellAbilityChoicesIterator.java

```diff
@@ -34,29 +34,42 @@ private static class ChoicePoint {
         Card selectedChoice;
         Score bestScoreForChoice = new Score(Integer.MIN_VALUE);
     }
-    private ArrayList<ChoicePoint> choicePoints = new ArrayList<>();
+    private final ArrayList<ChoicePoint> choicePoints = new ArrayList<>();
     private int incrementedCpIndex = 0;
     private int cpIndex = -1;
 
     private int evalDepth;
+    // Maps from filtered mode indexes to original ones.
+    private List<Integer> modesMap;
 
     public SpellAbilityChoicesIterator(SimulationController controller) {
         this.controller = controller;
     }
 
     public List<AbilitySub> chooseModesForAbility(List<AbilitySub> choices, int min, int num, boolean allowRepeat) {
         if (modeIterator == null) {
-            // TODO: Need to skip modes that are invalid (e.g. targets don't exist)!
+            // Skip modes that don't have legal targets.
+            modesMap = new ArrayList<>();
+            int origIndex = -1;
+            for (AbilitySub sub : choices) {
+                origIndex++;
+                if (sub.usesTargeting() && !sub.hasLegalTargets()) {
```

Comment:
Not sure if it exists (yes), but this might would still miss those Charms where the targeting happens on a sub of the choice
You would probably need to use it more along the lines of `ComputerUtilAbility.isFullyTargetable`

Reply:
Do you mean something like `Decisive Denial`? If so, that is being handled by this PR and I have a test for it included. Or do you mean something else?

Reply:
Yea that example works and I'm saying that skipping is still incomplete because you're only looking at the "root" (so ignoring the `DBFight` effect in this case)...

Proof:
![image](https://user-images.githubusercontent.com/8506892/209384256-5c6aa8e6-4c47-429a-a56a-904e0496ecca.png)

Reply:
Ah, gotcha. I now switched to `ComputerUtilAbility.isFullyTargetable()` and added a test for this.